feat(perf-issue): introduce multiple groups to post process forwarder

barkbarkimashark opened the pull request:
With the introduction of Performance issues, the assumption that an event will create at most one Group must change. Errors will always correspond to a single Group, but transactions can support up to 10 different types of Groups being created to accommodate for the different detection methods for a single transaction.

To support `post_process_group` working for transactions that result in multiple Groups being created, we need to modify the `EventStream.insert` and `PostProcessForwarderWorker` to accept a datastructure like:
```
{
  1: {
    "is_new": False,
    "is_new_group_environment": False,
    "is_regression": False,
  },
  2: {
    "is_new": True,
    "is_new_group_environment": False,
    "is_regression": False,
  }
}
```
that describes the groups that were created and its associated task state.


Fixes ISP-13

barkbarkimashark committed:
first pass at introducing groups_state to EventStream.insert and
post_process_forwarder

barkbarkimashark commented:
Planning to add more to this PR, but wanna get some early PR feedback to make sure I'm on the right track before digging in too deeply.

barkbarkimashark committed:
fix typing

rjo100 reviewed (commented):
been working on the `post_process_group` stuff and looks good on my end. one Q: should it be `GroupStates` or `GroupsState`? feel like the tense flip flops a bit

barkbarkimashark commented:
> been working on the `post_process_group` stuff and looks good on my end. one Q: should it be `GroupStates` or `GroupsState`? feel like the tense flip flops a bit

Hmm, I see what you mean - maybe `GroupsStates`? 

The structure is a map of groups and its associated states. 

GroupStates sounds like 'a single group and its various states'
GroupsState sounds like 'multiple groups and its one state'
GroupsStates sounds like 'multiple groups and its multuple states'?

barkbarkimashark commented on src/sentry/eventstream/kafka/backend.py:
```diff
@@ -69,6 +71,9 @@ def encode_bool(value: Optional[bool]) -> str:
                 value = False
             return str(int(value))
 
+        def encode_dict(value: Mapping) -> str:
+            return json.dumps(value)
```
Not sure if we want to encode the dict in a more optimal format.

wedamija replied:
I think this should be ok. I guess since these are mostly bools we could optimize if necessary

rjo100 commented:
I would say `GroupStates`. Slightly ambiguous because it could be multiple GroupState objects or multiple State objects associated with a single Group. But to most reading/reviewing the code should be obvious/explicit

barkbarkimashark committed:
rename GroupsState to GroupStates
fix GroupStates typing

barkbarkimashark committed:
fix typing

wedamija commented on src/sentry/eventstream/base.py:
```diff
@@ -22,6 +32,15 @@ class ForwarderNotRequired(NotImplementedError):
     """
 
 
+class GroupState(TypedDict):
+    is_new: bool
+    is_regression: bool
+    is_new_group_environment: bool
```
I think we need `primary_hash` here as well. We might not include this for perf issues, but I want to send error issues through here eventually as well.

wedamija reviewed (commented):
I think this is a good direction in general, mostly just small bits of feedback

barkbarkimashark commented on src/sentry/eventstream/snuba.py:
```diff
@@ -157,32 +160,37 @@ def insert(
+        ]
+
+        if groups_state is not None:
+            extra_data.append({**groups_state})
```
Hmm, maybe I'm not totally understanding the `EventStream` inheritance hierarchy. It looks like the `SnubaEventStream` doesn't make use of the extra data, but `KafkaEventStream` does? 

Then the `PostProcessForwarderWorker` processes the inserted payload (including message and headers) from kafka, which attempts to parse from the headers for the appropriate task kwargs and but will fallback to the message body if the header data is invalid. 

https://github.com/getsentry/sentry/blob/e126fbee41ea50543f55df64deaeea969531e517/src/sentry/eventstream/kafka/postprocessworker.py#L43-L56

For completeness, it might makes sense to include the group_state data in the header and message but I don't mind removing the extra data in the message itself.

wedamija replied:
`SnuabEventStream` makes use of the extra data here: https://github.com/getsentry/sentry/blob/7e85e5de95a9604cfaa81339448ee9db2f4a083d/src/sentry/eventstream/snuba.py#L393

You're right that `KafkaEventStream` uses it here: https://github.com/getsentry/sentry/blob/7e85e5de95a9604cfaa81339448ee9db2f4a083d/src/sentry/eventstream/kafka/backend.py#L186. But that's just for writing to the message body. These params are really only useful for the post process forwarder, snuba doesn't care about them. I think it's actually safe to not modify the message body, that way we don't have any concerns about breaking anything in the snuba consumer.

Post process forwarder actually only uses the headers. For the sake of understanding, assume the `post-process-forwarder:kafka-headers` and `eventstream:kafka-headers` options are always true. The exception path was just there to handle the cutover and as a failsafe. I don't see any occurrences of the error there in sentry in the last 90 days.

The old paths that process messages are no longer used and haven't been for a while now. I'd actually like to remove them as part of this work (separate pr though).

lynnagara replied:
The header hack was needed until we find a better long term solution for scaling the post process forwarder. It's the header option that was originally planned to be removed, not the other way around.

wedamija replied:
I didn't realise that the header thing was temporary. Tbh it seems like a good solution. Even if we scale this better using arroyo/remove semantic partitioning/etc, this results these consumers doing a lot less work overall

barkbarkimashark replied:
Fair enough. I'll assume that submitting the extra GroupStates to the header should be enough and remove the extra params from the body.

lynnagara replied:
We had many long discussions about this in the past. Ultimately the team (search and storage) agreed that the header can considered an optimization but can NOT be entirely relied on for important data which must also be in the body as well. I'm not sure if we want to revisit this decision but I think I'd prefer to keep it the same way right now and put it in both places.

barkbarkimashark replied:
Should I include the extra data in the message as well? Right now, its only in the header.

barkbarkimashark committed:
PR feedback.

barkbarkimashark committed:
fix typing

fpacifici commented on src/sentry/eventstream/kafka/protocol.py:
```diff
@@ -129,6 +136,20 @@ def decode_bool(value: bytes) -> bool:
+    if dict_str is None:
+        return None
+
+    parsed = json.loads(dict_str)
```
So se the comment above. One of the reasons we used headers was to avoid decoding json in the post process forwarder. Now this is likely to be mostly small but we should do it only if we really need.

lynnagara commented on src/sentry/eventstream/kafka/backend.py:
```diff
@@ -69,6 +71,9 @@ def encode_bool(value: Optional[bool]) -> str:
                 value = False
             return str(int(value))
 
+        def encode_dict(value: Mapping) -> str:
```
Please fix the type

lynnagara commented on src/sentry/eventstream/kafka/protocol.py:
```diff
@@ -129,6 +136,20 @@ def decode_bool(value: bytes) -> bool:
     return bool(int(decode_str(value)))
 
 
+def decode_optional_dict_str(value: Optional[bytes]) -> Optional[Mapping[Any, Any]]:
```
Should the type be optional[str] not Any?

lynnagara commented on src/sentry/eventstream/kafka/protocol.py:
```diff
@@ -172,9 +193,18 @@ def get_task_kwargs_for_message_from_headers(
+                header_data["group_states"] = None
+            try:
+                group_states = decode_optional_dict_str(header_data["group_states"])
+            except ValueError:
```
How come you are catching and ignoring this error?

wedamija replied:
Just being overcautious so that we don't unintentionally break anything. We should log the error here though.

barkbarkimashark replied:
Would it make more sense to let the Exception bubble up instead of logging it?

wedamija replied:
I don't want to end up skipping post process if this breaks for some reason. I think that once we're happy that this is working well we can remove the try/except though.

lynnagara replied:
I think this try/except is dangerous because if it fails we won't even know about it. If it's supposed to be removed then please add logging to Sentry here and a TODO stating that it is supposed to be temporary for rollout only.

lynnagara replied:
Ignore the part about logging, i see you already added that now

lynnagara commented on src/sentry/eventstream/snuba.py:
```diff
@@ -157,32 +161,37 @@ def insert(
+        extra_data: List[Mapping[Any, Any]] = [
+            {
+                "group_id": event.group_id,
+                "group_ids": [group.id for group in event.groups],
```
Curious why you're keeping the group ids in here and the group states relating to these groups in another place? Did you consider keeping it all together?

barkbarkimashark replied:
@wedamija it looks like `group_ids` in the message body isn't used? Can we safely delete this?

wedamija replied:
This is definitely used in the snuba consumer to write these ids to clickhouse. We can consolidate this into a dict if we want though, but you will need to make changes to the snuba consumer as well https://github.com/getsentry/snuba/blob/ea526f5af8b8405c814d47592793081c1c767b70/snuba/datasets/transactions_processor.py#L139-L141

wedamija replied:
Oh wait, this is in the message body right? If we're going to store the group ids in the headers then we don't really need to consolidate

barkbarkimashark replied:
Ahh, I see so `group_ids` is used in snuba to store to clickhouse. But the new `group_states` in the header will only be used in `post_process_forwarder`.

lynnagara commented on src/sentry/eventstream/base.py:
```diff
@@ -22,6 +32,16 @@ class ForwarderNotRequired(NotImplementedError):
+    primary_hash: Optional[str]
+
+
+GroupStates = Mapping[int, GroupState]
```
What's the int in the key going to be? Is that meant to be the group ID?

barkbarkimashark replied:
The int key is the group_id.

I was thinking about this and maybe it makes more sense to have something like this:

```
class GroupState(TypedDict):
  id: int
  is_new: bool
  is_regression: bool
  is_new_group_environment: bool
  primary_hash: Optional[str]

GroupStates = Sequence[GroupState]
```

lynnagara replied:
This seems more straightforward to me than keeping the sequence of group IDs in one place and the mapping of ID to the other data in another.

barkbarkimashark replied:
Not sure if this will be useful in the short term, but another positive is the list preserves the ordering.

Will make the change.

lynnagara commented on src/sentry/eventstream/base.py:
```diff
@@ -22,6 +32,16 @@ class ForwarderNotRequired(NotImplementedError):
+    is_new: bool
+    is_regression: bool
+    is_new_group_environment: bool
+    primary_hash: Optional[str]
```
How come it's optional? Shouldn't any group have a manatory primary hash?

wedamija replied:
We don't really need the primary hash for performance issues. It's only really used here https://github.com/getsentry/sentry/blob/1c5976ce038f60219d50a04647bb4a4128184ccf/src/sentry/tasks/post_process.py#L501-L507, and actually, all of the handlers just ignore it, so maybe we should just remove this altogether.

lynnagara commented on src/sentry/eventstream/snuba.py:
```diff
@@ -157,32 +161,37 @@ def insert(
+                "primary_hash": primary_hash,
+                "retention_days": retention_days,
+            },
+            {
+                "is_new": is_new,
+                "is_regression": is_regression,
+                "is_new_group_environment": is_new_group_environment,
+                "skip_consume": skip_consume,
+            },
```
What's the relationship between this and the new group states? Seems like we have some duplication now

wedamija replied:
We'll consolidate this together later on. Ideally I'd like to remove the individual group properties and just rely on groups instead.

This is also just in the body and is currently unused since we use the headers

barkbarkimashark replied:
Yeah, I'll open another PR to remove the redundant singular group params. Just wanted to slip in this multiple groups change and not make too many changes.

lynnagara replied:
Can you mark the parts that are supposed to be deprecated with comments or in the docstring then. Otherwise it can be hard to tell what direction this is supposed to be going in.

wedamija commented on src/sentry/eventstream/kafka/backend.py:
```diff
@@ -93,6 +98,7 @@ def strip_none_values(value: Mapping[str, Optional[str]]) -> Mapping[str, str]:
                     "is_regression": encode_bool(is_regression),
                     "skip_consume": encode_bool(skip_consume),
                     "transaction_forwarder": encode_bool(transaction_forwarder),
+                    "group_states": encode_dict(group_states) if group_states is not None else None,
```
The state about the group isn't stored in `rc-processing` unfortunately - `is_new`, `is_regression`, `is_new_group_environment` are all calculated while saving the event, and are currently passed via the headers. We could potentially store this data related to the group in a similar place and avoid passing it, that's just different to how we already handle this.

barkbarkimashark committed:
PR feedback

wedamija commented on src/sentry/event_manager.py:
```diff
@@ -1078,14 +1078,20 @@ def _eventstream_insert_many(jobs):
-            is_new = group_info.is_new
-            is_regression = group_info.is_regression
-            is_new_group_environment = group_info.is_new_group_environment
+            group_states = {
+                g.id: {
+                    "is_new": g.is_new,
+                    "is_regression": g.is_regression,
+                    "is_new_group_environment": g.is_new_group_environment,
+                }
+                for g in job["groups"]
+                if g is not None
+            }
```
We still need to support the old `is_new, is_regression` etc params for error issues

barkbarkimashark replied:
Yeah .. I complete misunderstood what was in `job["groups"]`.

barkbarkimashark committed:
correctly handle GroupInfo data

barkbarkimashark committed:
fix test

wedamija commented on src/sentry/eventstream/kafka/protocol.py:
```diff
@@ -172,9 +190,27 @@ def get_task_kwargs_for_message_from_headers(
+
+            group_states_str = None
+            try:
+                group_states_str = decode_optional_str(header_data["group_states"])
```
Does it makes sense to use `decode_optional_str(header_data.get("group_states"))` and skip the earlier check and mutating the header data?

wedamija commented on src/sentry/eventstream/kafka/protocol.py:
```diff
@@ -172,9 +190,27 @@ def get_task_kwargs_for_message_from_headers(
+            if "group_states" not in header_data:
+                header_data["group_states"] = None
+
+            group_states_str = None
```
Nit: We don't need this here, but we could have this as `group_states = None` and remove that same line from the except blocks.

wedamija commented on src/sentry/eventstream/kafka/protocol.py:
```diff
@@ -172,9 +190,27 @@ def get_task_kwargs_for_message_from_headers(
                 "is_regression": is_regression,
                 "is_new_group_environment": is_new_group_environment,
             }
```
Actually wondering if `group_states` should be included here instead? nbd either way

barkbarkimashark replied:
It might make sense to put it in there since it is apart of 'task state'. Are we planning to keep the flattened out params `is_new`, `is_regression`, and `is_new_group_environment` that are used for error issues as well as the multiple group states all in the task signature? 

If we keep `group_states` separate from `task_state` it'll be easier to do cleanup.

barkbarkimashark committed:
PR feedback

barkbarkimashark committed:
fix test

barkbarkimashark committed:
remove 'eventstream:kafka-headers' option usage

barkbarkimashark commented on src/sentry/eventstream/kafka/backend.py:
```diff
@@ -69,44 +70,29 @@ def encode_bool(value: Optional[bool]) -> str:
         def strip_none_values(value: Mapping[str, Optional[str]]) -> Mapping[str, str]:
             return {key: value for key, value in value.items() if value is not None}
 
-        # transaction_forwarder header is not sent if option "eventstream:kafka-headers"
-        # is not set to avoid increasing consumer lag on shared events topic.
-        transaction_forwarder = self._is_transaction_event(event)
-
-        send_new_headers = options.get("eventstream:kafka-headers")
```
Is this safe to delete? I looked for references and the only place this value is defined is in `src/sentry/options/defaults.py` and it just defaults to True. In what scenario would this not be True?

wedamija replied:
Based on what @lynnagara said I'm not totally sure if we should delete it without some more discussion. I'm in favour of deleting these paths, but others might disagree long term.

The option should be safe to delete. It is also set in the `sentry_options` table. I double checked and the value is `True` there as well

lynnagara replied:
Please don't delete this yet even though it may not be used. It was not supposed to be the default. We had to flip it because we were having big problems scaling single tenant.

lynnagara commented:
Can you run the typechecker over all your changes? I'm not sure if all the annotations are correct.

wedamija reviewed (approved):
This looks good to me in general. Let's deploy this tomorrow, and make the deploy isolated so we can watch this go out and roll back quickly if anything goes wrong.

barkbarkimashark commented:
> Can you run the typechecker over all your changes? I'm not sure if all the annotations are correct.

Is there any bit of code in particular you're worried about? 

The CI should automatically run mypy to do the type checking. 

I also just ran `mypy --strict --warn-unreachable` locally and it found no issues.

lynnagara reviewed (dismissed):
Please replace the `eventstream:kafka-headers` code that was removed and add sentry logging instead of ignoring failures to to parse the group states dict

barkbarkimashark committed:
Revert "remove 'eventstream:kafka-headers' option usage"

This reverts commit b13cdaa2635950230880ec435c2434d29615516d.

barkbarkimashark committed:
change structure of GroupState to include id
update GroupStates to be a sequence instead of a map
add todo to remove try/except

barkbarkimashark committed:
nest group_states within task_state
fix tests

barkbarkimashark commented on src/sentry/eventstream/kafka/protocol.py:
```diff
@@ -12,12 +13,15 @@ class UnexpectedOperation(Exception):
+        operation: str,
+        event_data: Mapping[str, Any],
+        task_state: Mapping[str, Any],
+        group_states: Optional[GroupStates] = None,
```
Yeah, decided to nest `group_states` into `task_state` - one fewer parameter

barkbarkimashark committed:
fix typing

barkbarkimashark committed:
forgot to change typing and encoding for GroupStates

barkbarkimashark committed:
include group_states data in message body

barkbarkimashark committed:
simplify

wedamija reviewed (approved):
lgtm, and it looks like @lynnagara's feedback is resolved.

barkbarkimashark committed:
fix test

barkbarkimashark committed:
fix message payload

barkbarkimashark committed:
fix test
